File: implementation1/bot_behavior/bot.py
from __future__ import annotations
from copy import deepcopy
from dataclasses import dataclass
from common_types import BotConfigInfo, BotMemoryInfo, BotPlayerInfo, BotState, ActionInfo, Action, PlayerAction, BombInfo, ExplosionInfo, WorldInfo, GridCoords
from .bot_state import AttackState, EscapeState, GetPowerupState, WanderState
import random


class BotController:
    def __init__(self, bot_config: BotConfigInfo):
        self._config = bot_config
        self._memory = BotMemory()
        self._current_state: BotState = WanderState()
        self._initialized = False
        self._last_bomb_count = 0
        self._last_explosion_count = 0

    @property
    def config(self) -> BotConfigInfo:
        return deepcopy(self._config)
    
    @property
    def memory(self) -> BotMemoryInfo:
        return deepcopy(self._memory)

    @property
    def current_state(self) -> BotState:
        return self._current_state

    def _initialize(self, dt: float, host_entity: BotPlayerInfo, world: WorldInfo):
            self._current_state.on_enter(self._config, self._memory, world, host_entity)
            self._initialized = True
            self._last_bomb_coords = {(b.row, b.col)
                                      for b in world.get_all_type(BombInfo)}
            self._last_explosion_coords = {
                (e.row, e.col) for e in world.get_all_type(ExplosionInfo)}
            
    def update(self, dt: float, host_entity: BotPlayerInfo, world: WorldInfo) -> None:
        # Init state on first frame
        if not self._initialized:
            self._initialize(dt, host_entity, world)

        should_force_reeval = False

        current_bomb_coords = {
            (b.row, b.col) for b in world.get_all_type(BombInfo)
        }
        current_explosion_coords = {
            (e.row, e.col) for e in world.get_all_type(ExplosionInfo)
        }

        should_force_reeval = False

        if self._memory.last_explosion_coords - current_explosion_coords:
            should_force_reeval = True

        new_bombs = current_bomb_coords - self._memory.last_bomb_coords
        for (r, c) in new_bombs:
            dist = max(abs(host_entity.col - c), abs(host_entity.row - r))
            if dist <= 5:
                should_force_reeval = True
                break

        self._memory.set_last_bomb_coords(current_bomb_coords)
        self._memory.set_last_explosion_coords(current_explosion_coords)


        timer_trigger = self._memory.tick_reeval(dt, self._config.reeval_interval, self._config.reeval_chance)
        is_in_danger = self._config.danger_check_type.is_in_danger(
            world, host_entity, self._config.danger_radius
        )

        # Check if danger!
        if is_in_danger:
            if not isinstance(self._current_state, EscapeState):
                self.transition_to(EscapeState(), world, host_entity)

        if timer_trigger or should_force_reeval or is_in_danger:
            self._perform_global_reevaluation(world, host_entity)

        if self._current_state:
            next_state = self._current_state.on_tick(self._config, self._memory, world, host_entity)
            if next_state:
                self.transition_to(next_state, world, host_entity)

    def decide_action(self, host_entity: BotPlayerInfo, world: WorldInfo) -> ActionInfo:
        action = self._current_state.decide_action(self._config, self._memory, world, host_entity)
        if action is None:
            return Action(action_type=PlayerAction.IDLE)
        return action

    def transition_to(self, new_state: BotState, world: WorldInfo, entity: BotPlayerInfo):
        # Clean up _memory before switching
        self._memory.set_path([])
        self._memory.set_goal(None)
        self._memory.set_strict_movement(False)

        self._current_state = new_state
        self._current_state.on_enter(self._config, self._memory, world, entity)

    def _perform_global_reevaluation(self, world: WorldInfo, bot: BotPlayerInfo):
        # 1 - Danger
        if self._config.danger_check_type.is_in_danger(world, bot, self._config.danger_radius):
            if not isinstance(self._current_state, EscapeState):
                self.transition_to(EscapeState(), world, bot)
            return

        # 2 - Powerup Check
        if random.random() <= self._config.powerup_chance:
            target = self._config.powerup_policy.get_goal(world, bot)
            path = self._config.powerup_policy.get_path(world, bot, self._memory)
            if target and path:
                if isinstance(self._current_state, GetPowerupState):
                    if self._memory.goal == target:
                        return

                self.transition_to(GetPowerupState(target), world, bot)
                return

        # 3 - Attack Check
        if isinstance(self._current_state, AttackState):
            return
        else:
            target_attack_pos = self._config.attack_policy.get_goal(world, bot)
            if target_attack_pos:
                self.transition_to(AttackState(target_attack_pos), world, bot)
                return

        # 4 - Default -> Wander
        if not isinstance(self._current_state, WanderState):
            self.transition_to(WanderState(), world, bot)


@dataclass
class BotMemory():
    """Memory for the bot."""

    def __init__(self):
        self._reeval_timer: float = 0.0

        # Navigation Memory
        self._path: list[GridCoords] = []
        self._goal: GridCoords | None = None
        self._is_strict_movement: bool = False

        # Perception _memory
        self._last_bomb_coords: set[GridCoords] = set()
        self._last_explosion_coords: set[GridCoords] = set()

    @property
    def reeval_timer(self) -> float:
        return self._reeval_timer

    def set_reeval_timer(self, value: float) -> None:
        self._reeval_timer = value

    @property
    def path(self) -> list[GridCoords]:
        return self._path

    def set_path(self, path: list[GridCoords]) -> None:
        self._path = path

    @property
    def is_strict_movement(self) -> bool:
        return self._is_strict_movement

    def set_strict_movement(self, value: bool) -> None:
        self._is_strict_movement = value

    @property
    def goal(self) -> GridCoords | None:
        return self._goal

    def set_goal(self, goal: GridCoords | None) -> None:
        self._goal = goal

    @property
    def last_bomb_coords(self) -> set[GridCoords]:
        return self._last_bomb_coords
    
    def set_last_bomb_coords(self, coords: set[GridCoords]) -> None:
        self._last_bomb_coords = coords

    @property
    def last_explosion_coords(self) -> set[GridCoords]:
        return self._last_explosion_coords

    def set_last_explosion_coords(self, coords: set[GridCoords]) -> None:
        self._last_explosion_coords = coords

    def tick_reeval(self, dt: float, reeval_interval: float, reeval_chance: float) -> bool:
        """Advance the reevaluation timer. Return `True` if reevaluation should occur."""
        self._reeval_timer += dt

        if self._reeval_timer >= reeval_interval:
            self._reeval_timer = 0.0
            return random.random() <= reeval_chance

        return False

File: implementation2/src/bot_behavior/bot.ts
import {
  BotInternalState,
  BotMemory,
  BotBehavior,
  WanderState,
  EscapeState,
  GetPowerupState,
  AttackState,
  HostileBot,
  CarefulBot,
  GreedyBot,
  Model,
  World,
  Player,
  GridCoords,
  Direction,
  Bomb,
  Explosion,
  BotType,
  BombOnlyDangerPolicy,
  AttackPolicy2,
  PowerupPolicy2,
  ExplosionPredictionDangerPolicy,
  AttackPolicy1,
  PowerupPolicy1,
  BotUpdateResult,
  BotConfig,
  IdleAction,
  Entity,
} from "../model";
import { Schema as S, HashSet, Match, Option } from "effect";

import * as Policies from "./helpers/policies";
import * as StateLogic from "./helpers/state";
import { getAllType } from "../helpers/world";


export const updateBot = (
  botState: BotInternalState,
  world: World,
  botEntity: Player,
  dt: number
): BotUpdateResult => {
  //  1 - INITIALIZATION
  let state = botState;
  let memory = state.memory;
  let currentBehavior = state.currentState;

  if (!state.initialized) {
    // Fill initial memory perception
    const bombs = getAllCoords(world, Bomb);
    const explosions = getAllCoords(world, Explosion);

    memory = {
      ...memory,
      lastBombCoords: bombs,
      lastExplosionCoords: explosions,
    };

    // Enter initial state
    const initRes = StateLogic.runOnEnter(
      currentBehavior,
      state.config,
      memory,
      world,
      botEntity
    );
    currentBehavior = initRes.state;
    memory = initRes.memory;

    state = { ...state, initialized: true };
  }

  // 2 - update PERCEPTION
  const currentBombs = getAllCoords(world, Bomb);
  const currentExplosions = getAllCoords(world, Explosion);
  let shouldForceReeval = false;

  const explosionDiff = HashSet.difference(
    memory.lastExplosionCoords,
    currentExplosions
  );
  if (HashSet.size(explosionDiff) > 0) {
    shouldForceReeval = true;
  }
  const newBombs = HashSet.difference(currentBombs, memory.lastBombCoords);
  if (HashSet.size(newBombs) > 0) {
    for (const b of newBombs) {
      const dist = Math.max(
        Math.abs(botEntity.row - b[0]),
        Math.abs(botEntity.col - b[1])
      );
      if (dist <= 5) {
        shouldForceReeval = true;
        break;
      }
    }
  }

  memory = {
    ...memory,
    lastBombCoords: currentBombs,
    lastExplosionCoords: currentExplosions,
  };

  // 3 - tick timer
  let timerTrigger = false;
  const newTimer = memory.reevalTimer + dt;

  if (newTimer >= state.config.reevalInterval) {
    memory = { ...memory, reevalTimer: 0 };
    if (Math.random() <= state.config.reevalChance) {
      timerTrigger = true;
    }
  } else {
    memory = { ...memory, reevalTimer: newTimer };
  }

  // 4 - check danger
  const isInDanger = Policies.isInDanger(
    state.config.dangerPolicy,
    world,
    botEntity,
    state.config.dangerRadius
  );

  if (isInDanger) {
    if (currentBehavior._tag !== "Escape State") {
      const res = transitionTo(
        EscapeState.make({ leftDanger: false }),
        state.config,
        memory,
        world,
        botEntity
      );
      currentBehavior = res.state;
      memory = res.memory;
    }
  }

  // 5 - reevaluate if we should
  if (timerTrigger || shouldForceReeval) {
    const res = performGlobalReevaluation(
      state.config,
      memory,
      currentBehavior,
      world,
      botEntity
    );
    currentBehavior = res.state;
    memory = res.memory;
  }

  // 6 - Tick the STATE
  const tickRes = StateLogic.runOnTick(
    currentBehavior,
    state.config,
    memory,
    world,
    botEntity
  );

  if (tickRes.nextState) {
      // Hard transition
      const res = transitionTo(
        tickRes.nextState,
        state.config,
        tickRes.memory,
        world,
        botEntity
      );
      currentBehavior = res.state;
      memory = res.memory;
  } else {
    memory = tickRes.memory;
  }

  // 7 - decide ACTION
  const decision = StateLogic.decideAction(
    currentBehavior,
    state.config,
    memory,
    world,
    botEntity
  );
  memory = decision.memory;

  return {
    nextState: {
      ...state,
      currentState: currentBehavior,
      memory: memory,
    },
    action: decision.action,
  };
};

// Helper to transition states (clean memory -> on_enter)
const transitionTo = (
  newState: BotBehavior,
  config: BotConfig,
  memory: BotMemory,
  world: World,
  bot: Player
) => {
  // Clean memory
  const freshMem = BotMemory.make({
    ...memory,
    path: [],
    goal: Option.none(),
    isStrictMovement: false,
    reevalTimer: 0
  });

  return StateLogic.runOnEnter(newState, config, freshMem, world, bot);
};

const performGlobalReevaluation = (
  config: BotConfig,
  memory: BotMemory,
  currentState: BotBehavior,
  world: World,
  bot: Player
) => {
  // 1 - Danger Check
  if (
    Policies.isInDanger(config.dangerPolicy, world, bot, config.dangerRadius)
  ) {
    if (currentState._tag !== "Escape State") {
      return transitionTo(
        EscapeState.make({ leftDanger: false }),
        config,
        memory,
        world,
        bot
      );
    }
    return { state: currentState, memory };
  }

  // 2 - Powerup Check
  if (Math.random() <= config.powerupChance) {
    const target = Policies.getPowerupGoal(config.powerupPolicy, world, bot);

    if (target) {
      if (
        currentState._tag === "Get Powerup State" &&
        isSameCoord(currentState.target, target)
      ) {
        return { state: currentState, memory };
      }
      return transitionTo(
        GetPowerupState.make({ target: target }),
        config,
        memory,
        world,
        bot
      );
    }
  }

  // 3 - Attack Check
  const attackTarget = Policies.getAttackGoal(config.attackPolicy, world, bot);
  if (attackTarget) {
    return transitionTo(
      AttackState.make({ target: attackTarget }),
      config,
      memory,
      world,
      bot
    );
  }

  // 4 - Default -> Wander
  if (currentState._tag !== "Wander State") {
    return transitionTo(WanderState.make({}), config, memory, world, bot);
  }

  return { state: currentState, memory };
};

// Helpers

const getAllCoords = <A extends Entity>(
  world: World, 
  type: S.Schema<A, any>
): HashSet.HashSet<GridCoords> => {
  const entities = getAllType(world, type);
  let set = HashSet.empty<GridCoords>();
  for (const e of entities) {
    set = HashSet.add(set, [e.row, e.col]);
  }
  return set;
};

const isSameCoord = (a: GridCoords, b: GridCoords) =>
  a[0] === b[0] && a[1] === b[1];
